Fix: Enable no-console rule in eslint-config-eslint (refs #5188)

platinumazure opened the pull request:
Some assumptions I made:

1. Makefile.js can do whatever it wants with the console
1. lib/logging.js can do whatever it wants with the console
1. lib/timing.js can do whatever it wants with the console (BUT let me know if it should pull in lib/logging.js instead)
1. The remaining files (all of which were unit tests) can use sinon sandboxes to stub console methods and have them be reliably restored, which happens to skirt the rule

Let me know if anything seems fishy. Thanks!

platinumazure commented on tests/lib/config/config-initializer.js:
```diff
@@ -232,25 +232,25 @@ describe("configInitializer", function() {
 
-            beforeEach(function() {
-                console.log = origLog;
-            });
```
I noticed that this beforeEach was trying to undo the before method's stubbing of console.log. That could also just be done in the before method itself. I do so as a try/catch/finally to ensure it is restored in happy and error paths. I believe this makes the logic around console.log stubbing more localized, and therefore more readable. If anyone disagrees, please let me know how I should do this instead.

IanVS replied:
Nice!  I definitely struggled with this one, so if this is working and not printing the progress bar to the console, then I like it much better.  Thanks.

platinumazure replied:
I ran `npm test` and didn't see any strange output, is that sufficient? Or is there more I need to do to check this case out?

IanVS replied:
You could also try `node_modules/.bin/mocha
tests/lib/config/config-initializer.js`.

On Thu, Apr 14, 2016 at 2:20 PM Kevin Partington <notifications@github.com>
wrote:

> In tests/lib/config/config-initializer.js
> <https://github.com/eslint/eslint/pull/5849#discussion_r59765989>:
>
> >                  }
> >              });
> >
> > -            beforeEach(function() {
> > -                console.log = origLog;
> > -            });
>
> I ran npm test and didn't see any strange output, is that sufficient? Or
> is there more I need to do to check this case out?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/eslint/eslint/pull/5849/files/815ef6069c646a90e3202c6d71d627ac2c14c771#r59765989>
>

platinumazure replied:
Can do. (And will do, tonight US time)

platinumazure commented:
Pinging:

@IanVS: I believe you did auto-config? Please take a look at the changes I made in tests/lib/config/config-initializer and let me know if you feel I have adversely affected the setup there.

@nzakas: Could you please review the assumptions I made in the initial comment of this pull request and let me know if I got anything wrong?

Thanks!

ilyavolodin commented:
I think the assumption made in timing is incorrect. Timing should output to console online in one place. Anything else is probably a mistake.

platinumazure commented:
@ilyavolodin How do you want that handled? Should I pull in logging, or should I find whoever is consuming lib/timing.js and put the console.log in the consuming code (and refactor lib/timing.js to return timing information in-memory instead of logging it)?

ilyavolodin commented:
@platinumazure timing is only used during development, so I don't think there's a need to refactor it. Instead of disabling no-console on the whole file, just disable it around one place where it's called. That way, if somebody adds another one, they will get flagged.

platinumazure commented:
@ilyavolodin Ah, so you're saying let's allow the one `console.log` but flag others. I can do that. Sorry for misunderstanding your original comment.

I'll update the branch tonight (US time).

platinumazure commented:
@ilyavolodin I've implemented your recommendation for lib/timing.js-- please review at your convenience and let me know if there are any outstanding problems there.

@IanVS Thank you for calling my attention to `node ./node_modules/mocha/bin/_mocha ./tests/lib/config/config-initializer.js`-- unfortunately it seems my change introduces a regression there. I will investigate and fix this.

Still not ready to merge, due to the above issue in the config-initializer tests.

platinumazure commented:
@IanVS Sorry for the flurry of messages. I just checked out HEAD^1 on my branch and I actually see logging statements even before my commit.

Would you mind trying to check out `git merge-base master platinumazure/no-console` (31bfe5f2ed) and see if your observations agree?

If so, I'm of two minds on fixing it-- my PR is not really intended to plug any gaps with console logging in test runs.

IanVS commented:
@platinumazure it looks fine to me when I run the test using your commit.  It does briefly show the progress bar, but it has always done that and I'm not worried about it.  The final output looks like:

```
  configInitializer
    processAnswers()
      prompt
        ✓ should create default config
        ✓ should disable semi
        ✓ should enable jsx flag
        ✓ should enable react plugin
        ✓ should not enable es6
        ✓ should extend eslint:recommended
        ✓ should not use commonjs by default
        ✓ should use commonjs when set
      guide
        ✓ should support the google style guide
        ✓ should support the airbnb style guide
        ✓ should support the standard style guide
        ✓ should throw when encountering an unsupported style guide
        ✓ should install required sharable config
      auto
        ✓ should create a config
        ✓ should create the config based on examined files
        ✓ should extend and not disable recommended rules
        ✓ should throw on fatal parsing error
        ✓ should throw if no files are matched from patterns


  18 passing (791ms)
```

So I'm 👍.  Good work.

ilyavolodin commented:
LGTM

platinumazure commented:
I've addressed all the feedback so far, so I am removing the "do not merge" label. I believe this is ready to be merged, but please speak up if you disagree.

platinumazure commented:
@nzakas Should this wait for a new "Chore" label? If so, should we set this as some combination of "blocked"/"do not merge"?

nzakas commented:
Yeah, let's wait for the Chore label.

gyandeeps commented:
@platinumazure I think you should be good now to update the label on this, then I can merge.

platinumazure commented:
Thanks @gyandeeps! I will do so in the next hour.
On Apr 22, 2016 8:00 PM, "Gyandeep Singh" <notifications@github.com> wrote:

> @platinumazure <https://github.com/platinumazure> I think you should be
> good now to update the label on this, then I can merge.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/eslint/eslint/pull/5849#issuecomment-213641744>
>

platinumazure commented:
Removing "do not merge" as I believe this is ready for review and merge once more.

gyandeeps commented:
LGTM :+1: 
Thanks for contributing.
